src/audio/multiband_drc/multiband_drc.c

```diff
@@ -94,94 +95,92 @@ static int multiband_drc_eq_init_coef_ch(struct sof_eq_iir_biquad *coef,
 	return 0;
 }
 
+/**
+ * @Description Initialize coefficients for multiband DRC processing.
+ *
+ * Allocates and initializes filter coefficients for the multiband DRC module. Memory
+ * for the crossover, emphasis, and de-emphasis filter coefficients is allocated within
+ * a contiguous block if not previously done. The function checks for configuration
+ * validity and adheres to predefined channel and band count limits.
+ *
+ * The memory layout for the coefficients_block is as follows:
+ * @code
+ * +-----------------------------------+
+ * |        coefficients_block         |
+ * +-----------------------------------+
+ * |       crossover_coef              | num_bands * nch * sizeof(struct sof_eq_iir_biquad)
+ * +-----------------------------------+ <-- offset for emp_coef (crossover_coef + num_bands * nch)
+ * |       emp_coef                    | num_bands * nch * sizeof(struct sof_eq_iir_biquad)
+ * +-----------------------------------+ <-- offset for deemp_coef (emp_coef + num_bands * nch)
+ * |       deemp_coef                  | num_bands * nch * sizeof(struct sof_eq_iir_biquad)
+ * +-----------------------------------+
+ * @endcode
+ *
+ * @parameters mod Pointer to the processing module containing multiband DRC data.
+ * @parameters nch Number of channels to process, up to PLATFORM_MAX_CHANNELS.
+ * @parameters rate Sampling rate, not used in current implementation.
+ *
+ * @return 0 on success or a negative error code on failure (e.g., invalid configuration,
+ *         memory allocation failure).
+ */
 static int multiband_drc_init_coef(struct processing_module *mod, int16_t nch, uint32_t rate)
 {
 	struct comp_dev *dev = mod->dev;
 	struct multiband_drc_comp_data *cd = module_get_private_data(mod);
-	struct sof_eq_iir_biquad *crossover;
-	struct sof_eq_iir_biquad *emphasis;
-	struct sof_eq_iir_biquad *deemphasis;
 	struct sof_multiband_drc_config *config = cd->config;
 	struct multiband_drc_state *state = &cd->state;
 	uint32_t sample_bytes = get_sample_bytes(cd->source_format);
-	int i, ch, ret, num_bands;
+	struct sof_eq_iir_biquad *crossover, *emphasis, *deemphasis;
+	int i, ch, ret;
+	int num_bands = config ? config->num_bands : 0;
```

Comment:
no need to check for `config` twice. The original code, checking `config != NULL` in line 137 below and then assigning `num_bands` unconditionally was fine. This change isn't an improvement IMHO.